### ISSUE TITLE
[V3] v2 -> v3 data migration

### ISSUE THREAD

**Issue by d-v-b** [zarr-developers/zarr-python#1798]
We should invest in tools to make the v2 -> v3 conversion simple for people who are motivated to convert their data. A few high-level ideas:
- A simple CLI that converts an array, or a group (recursive or not) from v2 to v3, in a new location.
- - ~~Someone should investigate how complicated in-place conversions would be. On a local filesystem where `mv` is cheap, this could be attractive.~~. V3 is designed to make array conversions easy, requiring only the creation of new metadata.
- - The CLI should use functions that are accessible from scripts that don't use the CLI. We can look at work @normanrz did in [Zarrita](https://github.com/scalableminds/zarrita/blob/b58c257221f012621f9084c9e1be8d526d8b34eb/zarrita/array_v2.py#L452).
- Documentation of the key differences between `zarr-python` v2 and v3, and a migration guide. This should have its own page in the docs.
- - We should consider options for people who don't want to re-save their data. I'm not presently a kerchunk user, but I presume that kerchunk could map v2 to v3, for people who don't want to convert their data? cc @martindurant.

**Comment by jeromekelleher** [comment]
Big +1 on this. I'm working on a conversion tool for large-scale genomics data (100s TB scale) which is usually held in file systems (for the moment, it will probably migrate to object stores later on). A CLI tool that does an in-place migration from v2 to v3 would be a big help. I'm hoping to move the v3 early on, before too many datasets are converted into v2 format and so most users won't ever know about v2.

My assumptions was that the migration was largely a case of writing a new JSON metadata file per-array, and should be possible to do both cheaply and safely?

**Comment by d-v-b** [comment]
> My assumptions was that the migration was largely a case of writing a new JSON metadata file per-array, and should be possible to do both cheaply and safely?

Yes, I think this is right. Besides the metadata, which will live in a completely new JSON document (`zarr.json`), V3 supports a [backwards-compatible layout for the chunks](https://github.com/zarr-developers/zarr-specs/blob/81317acfc1441ff4274bcfbf67386e9852892969/docs/v3/core/v3.0.rst?plain=1#L543)

**Comment by jeromekelleher** [comment]
Thanks yes, I've been aiming for v3 forwards compatibility by using "/" as the default dimension separator.  Then, iterating over the chunks in the first dimension and renaming to have a "c" prefix should be relatively cheap (I forgot about this difference).

Is there some developer documentation with recommendations for forwards/backwards compatibility?

**Comment by normanrz** [comment]
> * Someone should investigate how complicated in-place conversions would be. On a local filesystem where `mv` is cheap, this could be attractive.

For most cases, the migration only requires adding `zarr.json` files throughout the hierarchy. There should be no need to even touch the chunk files. `zarr.json` and `.zarray` files can also live side-by-side. So, why would a `mv` be needed?
Only, when using a non-supported codec or filter, chunks need to be rewritten.

**Comment by d-v-b** [comment]
> So, why would a mv be needed?
Only, when using a non-supported codec or filter, chunks need to be rewritten.

This is correct. When I wrote up this issue, I forgot about the `v2` chunk key encoding supported by v3 🤦

**Comment by d-v-b** [comment]
i updated the issue to be more accurate :)

**Comment by normanrz** [comment]
I agree that a CLI tool that can convert an entire hierarchy would be great!

**Comment by jhamman** [comment]
Today I learned that there is a v1 to v2 migrator in the zarr-python codebase: https://github.com/zarr-developers/zarr-python/blob/6105ef203e3e5c5390aa01db613bc42e8fb98c1a/zarr/storage.py#L1941-L1956

**Comment by jhamman** [comment]
#2596 was mislabeled as closing this. A migration tool would still be great!

**Comment by meteoDaniel** [comment]
I am intersted, too. 

Without the opportunity to migrate Data from v2 to v3, it would be a nightmare to deprecate zarr v2 support. 
M

**Comment by meteoDaniel** [comment]
@eschechter thanks for sharing! 

So it is enough to adapt the metadata without touching the data itself?

**Comment by eschechter** [comment]
That's correct @meteoDaniel - the zarr v3 conversion can be done with just the creation of zarr.json metadata files. No need to touch the data.

**Comment by K-Meech** [comment]
Hi all, I'd be happy to put together a draft PR of a v2 > v3 conversion CLI (expanding / extending [@eschechter 's gist](https://github.com/zarr-developers/zarr-python/issues/1798#issuecomment-2725207137) - which is a great start :smile:). Do let me know if anyone is already working on this (happy to collaborate!) + if there are any preferences for structure / location of the CLI in the existing codebase.

**Comment by normanrz** [comment]
That's great! I don't think there are any CLI tools in this codebase already. So, I guess we're open for suggestions on the structure.
I think it would be great, if the conversion functions could also be called from Python directly. A CLI would then be using these functions.
I had good experiences with building CLIs with typer. That should probably go in a new dependency group, e.g. `pip install zarr[cli]`.
I think the CLI should only attempt to do metadata-only conversions. If that is not possible (e.g. due to missing codecs in v3) it should fail.

**Comment by d-v-b** [comment]
with the CLI added and improved documentation, I think this can be closed as completed.
